File: Data Structures & Algorithms/two-integer-sum/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[]}
     */
    twoSum(nums, target) {
        const map = new Map();

        for (let i = 0; i < nums.length; ++i) {
            map[nums[i]] = i
        }

        for (let i = 0; i < nums.length; ++i) {
            let j = map[target - nums[i]]
            if (j && j != i) {
                return [i, j]
            }
        }

        return [0, 0]
    }
}

File: Data Structures & Algorithms/target-sum/submission-2.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        memo = {}

        def dp(i, remain):
            if (i, remain) in memo:
                return memo[(i, remain)]

            if i >= len(nums):
                return remain == target

            memo[(i, remain)] = dp(i + 1, remain + nums[i]) + dp(i + 1, remain - nums[i])

            return memo[(i, remain)]

        return dp(0, 0)